### PR TITLE
Emojis

### PR DESCRIPTION
**Pull request by 4z0t**
Adds emojis in game chat and lobby chat
still required mod to access Packs for user emojis

edit by Jip:
 - [ ] Moderators confirmed emoticons are good to go.

### CODE REVIEW

**Review by speed2CZ — CHANGES_REQUESTED** [review]
Please put back table.empty() calls, that's the fastest way to find out if the teable is empty
Can you cleanup all the useless comments?

**Comment by 4z0t** [comment]
> Please put back table.empty() calls, that's the fastest way to find out if the teable is empty
> Can you cleanup all the useless comments?

Done

**Comment by KionX** [comment]
The code looks dirty. Trailing spaces, lots of empty lines, confused tabulation.
It looks like the PR is based on old code and can undo other changes.
Also, this PR is not self-sufficient. Should consider the option without mods.

**Comment by 4z0t** [comment]
Any problems left? I think its done, and im awaiting your decision.

**Comment by speed2CZ** [comment]
Well I look at the first file lua/ui/game/chat.lua and it has tons of commented out code.

**Comment by 4z0t** [comment]
Seems to better now 🙂

**Comment by 4z0t** [comment]
> Well I look at the first file lua/ui/game/chat.lua and it has tons of commented out code.
So, I think its about to be done, what do you think?

**Comment by Uveso** [comment]
@4z0t 

How can i test this ?

**Comment by 4z0t** [comment]
@Uveso i can send u some emojis to test

**Review by Uveso — APPROVED** [review]
Tested; also in multiplayer. Works good.
table.empty are also back.

File: lua/ui/lobby/chatarea.lua [hunk 1]
```diff
@@ -16,19 +21,19 @@ ChatArea = Class(Group){
         -- Each line can have multiple TextFields and each TextField can have its own style
         -- All chat lines are indexed by the order in which they were created - so older messages are first
         self.ChatLines = {} -- for example:
-        -- ChatLines = {
-        --   { TextFields = { { text = "Mike" }, { text = "Hi Dave, How are you?" } } },
-        --   { TextFields = { { text = "Dave" }, { text = "I'm fine, how are you" } } }
-        --}
+        -- ChatLines = {
```
Comment (speed2CZ) [inline lua/ui/lobby/chatarea.lua:24]: Why all this new whitespace?

File: lua/ui/lobby/chatarea.lua [hunk 2]
```diff
@@ -49,30 +54,30 @@ ChatArea = Class(Group){
             LOG('chatArea dirty Height')
             self:ReflowLines()
         end
-
+         
         LayoutHelpers.AtLeftTopIn(self, parent, 0, 0)
     end,
-
+
```
Comment (speed2CZ) [inline lua/ui/lobby/chatarea.lua:60]: and here

File: lua/ui/lobby/chatarea.lua [hunk 3]
```diff
@@ -49,30 +54,30 @@ ChatArea = Class(Group){
...
             lineSpacing = 1, -- don't use higher values because chat looks strange with small fonts
             padding = { left = 0, top = 3, right = 3, bottom = 3 },
         }
     end,
-
+    
     --- posts a new chat message and its author name using optional styling tables
     PostMessage = function(self, messageText, authorName, messageStyle, authorStyle)
-
+
```
Comment (speed2CZ) [inline lua/ui/lobby/chatarea.lua:76]: and so on

File: lua/ui/lobby/chatarea.lua [hunk 4]
```diff
@@ -99,11 +104,13 @@ ChatArea = Class(Group){
...
-        authorStyle.padding.left = 0
-        authorStyle.padding.right = 0
+        authorStyle.padding.left = 0 
+        authorStyle.padding.right = 0 
 
-        -- always use bolder font for the Author name
+        -- always use bolder font for the Author name 
+       
 
+
```
Comment (speed2CZ) [inline lua/ui/lobby/chatarea.lua:113]: two extra lines for nothing?

File: lua/ui/lobby/chatarea.lua [hunk 5]
```diff
@@ -114,97 +121,125 @@ ChatArea = Class(Group){
...
-        -- If no author provided, we're simply going to skip it. Else we're going to style it
+        
+
         if authorName == nil then
-           authorName = ''
-        else
-            authorName = '['..authorName..'] '
+            authorName = ''
+        else 
+            authorName = '['..authorName..'] '
```
Comment (speed2CZ) [inline lua/ui/lobby/chatarea.lua:129]: again spaces for nothing

File: lua/ui/lobby/chatarea.lua [hunk 6]
```diff
@@ -114,97 +121,125 @@ ChatArea = Class(Group){
...
-        -- wrap message text and its author into multiple lines based on control's width
-        local wrapLines = Text.WrapText(chatText,
-            self.Width() - self.Style.padding.left - self.Style.padding.right,
-            customAdvanceFunction)
+        local contents = self:WrapContents({
+                                                text = Emojis.CheckEmojis(messageText,Prefs.GetFromCurrentProfile("ChatLobbyEmojis")),
+                                                style = messageStyle
+                                            },
+                                            {
+                                                text = authorName,
```
Comment (speed2CZ) [inline lua/ui/lobby/chatarea.lua:139]: If you're using tabs, convert it to 4 spaces

File: lua/ui/lobby/chatarea.lua [hunk 7]
```diff
@@ -114,97 +121,125 @@ ChatArea = Class(Group){
...
-        else
+        else 
            -- putting the line below previous line
            line.previous = self.ChatLines[line.index - 1]
            line.Left:Set(function() return self.Left() end)
            LayoutHelpers.Below(line, line.previous)
         end
         -- store the new line so we can destroy all lines later
-        table.insert(self.ChatLines, line)
+        table.insert(self.ChatLines, line)
```
Comment (speed2CZ) [inline lua/ui/lobby/chatarea.lua:234]: again white spaces around this area

File: lua/ui/lobby/chatarea.lua [hunk 8]
```diff
@@ -294,12 +325,82 @@ ChatArea = Class(Group){
...
     OnDestroy = function(self)
         self:ClearLines()
         self.ChatLines = nil
         self.ChatHistory = nil
         Group.OnDestroy(self)
     end,
-}
\ No newline at end of file
+
+
```
Comment (speed2CZ) [inline lua/ui/lobby/chatarea.lua:337]: one empty line is enough

File: lua/ui/lobby/chatarea.lua [hunk 9]
```diff
@@ -294,12 +325,82 @@ ChatArea = Class(Group){
...
+                    return linewidth - 4
+                end
+            end, 
+            function(text)
+                return self:AdvanceFunction(text, contents.style)
+            end)
+    end,
+    
+    
+
```
Comment (speed2CZ) [inline lua/ui/lobby/chatarea.lua:353]: again just one  empty line please

File: lua/ui/lobby/emojis.lua [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+            pack.info.isEnabled = packsStates[packname]
+        else
+            packsStates[packname] = true
+            pack.info.isEnabled = true
+        end
+    end
+    Prefs.SetToCurrentProfile("emojipacks", packsStates)
+end
+
+do
```
Comment (speed2CZ) [inline lua/ui/lobby/emojis.lua:51]: you dont need "do end" here, the code is executed when the file is imported. So just the function call is enough

File: lua/ui/lobby/lobby.lua [hunk 11]
```diff
@@ -1227,7 +1235,7 @@ local function autobalance_bestworst(players, teams_arg)
     end
 
     -- teams first picks best player and then worst player, repeat
-    while not table.empty(players) do
+    while not table.empty(players)  do
```
Comment (speed2CZ) [inline lua/ui/lobby/lobby.lua:1238]: Alse clean this file of white spaces

**Review by speed2CZ — CHANGES_REQUESTED** [review]
There's still a lot of white spaces that needs to go. Clean it up please

**Comment by Uveso** [comment]
Tested (also multiplayer), works good.

File: lua/ui/lobby/lobby.lua [hunk 12]
```diff
@@ -5235,7 +5291,8 @@ function CPUBenchmark()
     local currTime
     local countTime = 0
     --Make everything a local variable
-    --This is necessary because we don't want LUA searching through the globals as part of the benchmark
+    --This is necessary because we don't want LUA searching through the globals as part of the benchmar
```
Comment (shalkya) [inline lua/ui/lobby/lobby.lua:5294]: benchark
smh the k got eaten up in this pr

File: lua/ui/lobby/lobby.lua [hunk 13]
```diff
@@ -5767,6 +5844,194 @@ function ShowLobbyOptionsDialog()
...
+        end
+        for ind = lineIndex, self.numLines do self.LineGroup.Lines[ind]:render() end
+    end
+
+      -- called to determine if the control is scrollable on a particular access. Must return true or false.
+    GUI.PackageManager.IsScrollable = function(self, axis)
+        return true
+    end
+
+    --scrlling
```
Comment (shalkya) [inline lua/ui/lobby/lobby.lua:5937]: scrolling

File: lua/ui/lobby/lobby.lua [hunk 14]
```diff
@@ -5767,6 +5844,194 @@ function ShowLobbyOptionsDialog()
...
+                    self.name:SetText('')
+                    self.author:SetText('')
+                    self.desc:Clear()
+                    self.bg:Disable()
+                end
+
+            end
+
+            line.bg.OnCheck = function(self, checked)
+                LOG('set '..repr(checked)..' on '..repr(self.id))
```
Comment (shalkya) [inline lua/ui/lobby/lobby.lua:6010]: LOG should be removed or at least commented

**Comment by shalkya** [comment]
alright so from what I tested, it doesn't work in the lobby at first, even if it's ticked on by default. You actually need to untick, tick it on again, to make it work.
I would be better to fix that, so people aren't confuse. Either make it unticked by default, so they have to tick it. Or fix the issue of it not working the first time. I think you said something about having no data in the gamepref.

**Comment by shalkya** [comment]
you still have the emoji completion on, even with the setting turn off in the lobby. It shouldn't display the completion, if that setting is off, like it's working in the game chat.

**Comment by 4z0t** [comment]
@shalkya yeah, those options are separated, that means one could be turned on, but another off

**Comment by shalkya** [comment]
i'm sorry but i don't see the setting to turn off to remove the completion in lobby for emojis. 
image
anyway (if there is several option like you are saying) the option for emojis showing in lobby and the one for completion of emojis in lobby should be the same.
But i think you didn't understand what i said before, and you were only commenting on the fact that there is one option for the lobby and another for the game. Which is not an issue.

**Review by speed2CZ — APPROVED** [review]

**Comment by Garanas** [comment]
If I am correct these changes are reviewed and approved. I've fixed the conflict / will fix future conflicts and I'll merge this on Monday morning. Great work everyone!

**Comment by Garanas** [comment]
[Packs.zip](https://github.com/FAForever/fa/files/7491211/Packs.zip)

Emoji pack provides by @4z0t to be stored here:

image

**Comment by Garanas** [comment]
Moved this PR to the next milestone. I intent to introduce a similar structure to the custom strategic icons where you can define a mod_emoticons.lua file that has a mapping of `key -> emotion`. In its current state other UI mods can not introduce emoticons.

**Comment by Sheikah45** [comment]
Shouldn't all emoticons supported be packaged with the base game? Since this is not something like strategic icons where only you will and are supposed to see it. If UI mods are the source for the emoticons then it will very easily end up being the case where one person thinks they are sending emoticons but the other person can only see text. And has no way of getting the emoticons unless they know exactly which UI mod to install. And even then there could be collisions between UI mods in terms of emoticon names.

**Comment by 4z0t** [comment]
@Sheikah45 all emojis have package name and then their in-package name like `faf/uef`, so collisions may be only if packages are the same.

**Comment by Garanas** [comment]
I'm fine if they become part of the base game, but then I propose we first make a selection and add it in before merging this.

**Comment by 4z0t** [comment]
If we are going to add them in base game, then mb we have to ask community what emojis to add, that I'll convert into dds.

**Comment by Garanas** [comment]
@4z0t Could you make a forum post about this? I'll get it into the news. Please make sure that you have the specifications (width, height, resolution - etc) and add that a submission doesn't mean it gets embedded - it merely means we'll consider it 😄 .

We can make this a bi-monthly thing - as embedding emoticons to the game is as easy as waiting for the next developers patch.

**Comment by Garanas** [comment]
Reminder to ping the moderators before we merge any emoticons:
 - [ ] Moderators confirmed emoticons are good to go.

**Comment by Garanas** [comment]
After discussing it with 4z0t, I am closing this pull request. The forum post on what emoticons to add didn't get enough traction. This pull request still requires quite a bit of work - as an example, if UI mod support is enabled then it needs a similar framework as the strategic icons loading has. But then during the lobby already.

**Pull request closed without merging**